### ISSUE TITLE
Cannot upload large file

### ISSUE THREAD

**Issue by marius-avram** [mscdex/busboy#143]
I'm trying to implement uploading a file to a server using node.js and busboy. 
I need however to transfer larger files (up to 10 GB). I'm testing with a 3.5GB locally and it doesn't work at all. After part of the upload is process the upload process starts over all over again. I think this is what it does as the progress returns back to zero.

I'm using the example code:

```
var http = require('http'),
    path = require('path'),
    os = require('os'),
    fs = require('fs');

var Busboy = require('busboy');

http.createServer(function(req, res) {
  if (req.method === 'POST') {
    var busboy = new Busboy({ headers: req.headers });
    busboy.on('file', function(fieldname, file, filename, encoding, mimetype) {
      file.pipe(fs.createWriteStream('file.out'));
    });
    busboy.on('finish', function() {
      res.writeHead(200, { 'Connection': 'close' });
      res.end("That's all folks!");
    });
    return req.pipe(busboy);
  }
  res.writeHead(404);
  res.end();
}).listen(8000, function() {
  console.log('Listening for requests');
});
```

If I don't pipe the content and I'm only listening for the data event the whole upload process is completed. However this is useless because I don't process the incoming chunks in any way.

```
var http = require('http'),
    inspect = require('util').inspect;

var Busboy = require('busboy');

http.createServer(function(req, res) {
  if (req.method === 'POST') {
    var busboy = new Busboy({ headers: req.headers });
    busboy.on('file', function(fieldname, file, filename, encoding, mimetype) {
      console.log('File [' + fieldname + ']: filename: ' + filename);
      file.on('data', function(data) {
        // Just do nothing
      });
      file.on('end', function() {
        console.log('File [' + fieldname + '] Finished');
      });
    });
    busboy.on('field', function(fieldname, val, fieldnameTruncated, valTruncated) {
      console.log('Field [' + fieldname + ']: value: ' + inspect(val));
    });
    busboy.on('finish', function() {
      console.log('Done parsing form!');
      res.writeHead(303, { Connection: 'close', Location: '/' });
      res.end();
    });
    req.pipe(busboy);
  } else if (req.method === 'GET') {
    res.writeHead(200, { Connection: 'close' });
    res.end('<html><head></head><body>\
               <form method="POST">\
                <input type="text" name="textfield"><br />\
                <select name="selectfield">\
                  <option value="1">1</option>\
                  <option value="10">10</option>\
                  <option value="100">100</option>\
                  <option value="9001">9001</option>\
                </select><br />\
                <input type="checkbox" name="checkfield">Node.js rules!<br />\
                <input type="submit">\
              </form>\
            </body></html>');
  }
}).listen(8000, function() {
  console.log('Listening for requests');
});
```

**Comment by mscdex** [comment]
Did you try using unique filenames with `fs.createWriteStream()`?

**Comment by marius-avram** [comment]
Yes. My code is more elaborate than the example. However I tried the example because I thought it should work. However it doesn't. After a while the upload is being reset. 

Also I'm trying to upload only one file at a time so I don't think the filename should be any problem. I think this is more of a buffer problem. And and the moment I'm not sure the library is the reason why it doesn't work.

I tested both on Chrome and on Firefox. The problem persists. I'm using Ubuntu 16.04 as a operating system. Maybe is from the operating system ?

Did you test the library with bigger files ? Like 2GB ? And are you able to write them on disk ?

**Comment by marius-avram** [comment]
As a temporary solution I chose to send the file in chunks of 50MB. I used dropzone + resumable.js in the frontend.  Had to rewrite a lot of my backend code, but it seems to work.

**Comment by mayeaux** [comment]
@marius-avram can you make a gist with that code?

**Comment by mscdex** [comment]
FWIW I still cannot reproduce this issue. I was able to upload a 4.3GB file without any problem.

**Comment by lanzosuarez** [comment]
@mscdex  can you show me how you are doing the writing of the file? 

This one, works, but it's broken, and the file size is much larger than the total
![image](https://user-images.githubusercontent.com/18522203/37012440-95b705aa-212f-11e8-976e-e1b6b1d38513.png)
`

Tried this one, create's a file too, but i think the last chunk wasnt writted, that's why the size is smaller than the actual
![image](https://user-images.githubusercontent.com/18522203/37012479-d3e26c02-212f-11e8-9913-b3a8244ed282.png)

Thanks in advacnce

**Comment by mscdex** [comment]
@lanzosuarez You don't `pipe()` within a `'data'` event handler, you do so inside the `'file'` event handler.

**Comment by lanzosuarez** [comment]
@mscdex yes thank you. I overlooked it. Thanks a lot

**Comment by radavi** [comment]
@mscdex: How are you sending the file from client to the server for the 4.3 GB? in chunks? We have a client_max_body_size set t0 1.5 GB and we can't transfer files which are larger than the limit. So we have to send them in chunks. However we are having some difficulty doing that. Can post a jsfiddle for it if you can? Thanks.

**Comment by mscdex** [comment]
@radavi I don't know what "client_max_body_size" is. For my test I submitted data from an html form directly to a plain node http server that parsed the request using `busboy`.

**Comment by radavi** [comment]
@mscdex: "client_max_body_size" is something that is set at the NGINX layer where you can limit the file size transferred from client to server. So, if you pass anything over the set limit, its going to be blocked at the transport layer. Having said that, if you can pass the file over from client to server in chunks, we don't have to hit the limit. Is there some sample code that you can share?

**Comment by mscdex** [comment]
@radavi If nginx's limit is what you're hitting then you'll have to do something creative client-side. As I said, my example was a direct connection between the browser and a plain node http server, meaning no middle man (e.g. nginx). There are js libraries that will chunk the data for you and then you just have to handle that appropriately on the server-side.

**Comment by vijayakumar-psg587** [comment]
@mscdex apologies to resurrect an old thread, but just wanted to know, if you have tried by simulating multiple requests, or a single request. 4.3GB is a huge file, so just wanted to get some details on how long did it take to upload ? I tried with a 439MB file and it took almost 5 seconds. So wanted to know if it can be made more efficient or this is expected scenario

**Comment by mscdex** [comment]
@vijayakumar-psg587 there are so many variables when it comes to that sort of thing. For example: node versions, OS, transfer over network/internet vs. localhost, how the client is transferring the data, what the server is doing with the data (e.g. saving to disk or transparently uploading to another server or something else), etc.

I did not measure the performance at the time I tested with that particular configuration.
